src/cluster_import.c

```diff
@@ -0,0 +1,1915 @@
+                  "(owner of slots [%s]). Failing import request.",
+                  link->linkname,
+                  link->nodename,
+                  link->slot_ranges_str);
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED, "Failover denied");
+        return;
+    }
+    serverLog(LL_WARNING,
+              "Slot import link %.40s had failover denied from node %.40s "
+              "(owner of slots [%s]). Going back to streaming incremental updates.",
+              link->linkname,
+              link->nodename,
+              link->slot_ranges_str);
+    updateSlotMigrationLinkStatusMessage(link, "Previous failover attempt denied from source");
+    updateSlotMigrationLinkState(link, SLOT_IMPORT_RECEIVE_STREAM);
+}
+
+/* Sent by the source to the target to notify of some unrecoverable error. */
+void clusterCommandSyncSlotsFail(client *c) {
+    if (!c->flag.slot_import_source || !c->slot_migration_link) {
+        serverLog(LL_WARNING,
+                  "Received CLUSTER SYNCSLOTS FAIL from client %llu, "
+                  "but the client is not a slot import source. Closing link.",
+                  (unsigned long long)c->id);
+        freeClientAsync(c);
+        return;
+    }
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
+    if (!isSlotMigrationLinkInProgress(link)) {
+        return;
+    }
+    finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED, "Source node terminated import");
+}
+
+slotMigrationLink *createSlotImportLink(clusterNode *source_node, list *slot_ranges, int one_shot) {
+    slotMigrationLink *link = zcalloc(sizeof(slotMigrationLink));
+    link->ctime = server.unixtime;
+    link->last_update = link->ctime;
+    link->last_ack = link->ctime;
+    link->type = SLOT_MIGRATION_IMPORT;
+    link->slot_ranges = slot_ranges;
+    link->slot_ranges_str = representSlotRangeList(slot_ranges, '-');
+    link->one_shot = one_shot;
+    getRandomHexChars(link->linkname, sizeof(link->linkname));
+    memcpy(link->nodename, source_node->name, CLUSTER_NAMELEN);
+    serverLog(LL_NOTICE,
+              "New slot import link created: link name %.40s, "
+              "source node %.40s, slot ranges %s.",
+              link->linkname,
+              link->nodename,
+              link->slot_ranges_str);
+    return link;
+}
+
+void slotImportConnectHandler(connection *conn) {
+    slotMigrationLink *link = (slotMigrationLink *)connGetPrivateData(conn);
+    proceedWithSlotMigration(link);
+}
+
+void connectSlotImportLink(slotMigrationLink *link, char *ip, int port) {
+    sds status_msg;
+    serverLog(LL_NOTICE,
+              "Connecting slot import link %.40s to %.40s "
+              "(owner of slots [%s]) at (ip: %s, port %d)",
+              link->linkname,
+              link->nodename,
+              link->slot_ranges_str,
+              ip,
+              port);
+
+    link->conn = connCreate(connTypeOfCluster());
```

Comment:
```suggestion
    link->conn = connCreate(connTypeOfCluster());
```
Presumably this should be the replication type, since we're connecting to the default port, which might be different. (IE, cluster might be TCP but replication is TLS)

Reply:
Right. Yeah we want to reuse tls-replication, not tls-cluster setting here